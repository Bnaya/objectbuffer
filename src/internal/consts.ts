export const GLOBAL_LOCK_OFFSET = 0;
export const INITIAL_ENTRY_POINTER_TO_POINTER =
  GLOBAL_LOCK_OFFSET + Uint32Array.BYTES_PER_ELEMENT;

export const ENDIANNESS_FLAG_POINTER =
  INITIAL_ENTRY_POINTER_TO_POINTER + Uint32Array.BYTES_PER_ELEMENT;

export enum ENDIANNESS {
  BIG,
  LITTLE,
}

export const TEMP_SAVE_POINTER =
  ENDIANNESS_FLAG_POINTER + Uint32Array.BYTES_PER_ELEMENT;

export const INITIAL_ENTRY_POINTER_VALUE =
  TEMP_SAVE_POINTER + Uint32Array.BYTES_PER_ELEMENT;

export const MEM_POOL_START =
  INITIAL_ENTRY_POINTER_VALUE + Uint32Array.BYTES_PER_ELEMENT;

export const UNDEFINED_KNOWN_ADDRESS = 0;
export const NULL_KNOWN_ADDRESS = 1;
export const TRUE_KNOWN_ADDRESS = 2;
export const FALSE_KNOWN_ADDRESS = 3;
export const MAX_64_BIG_INT = BigInt("0xFFFFFFFFFFFFFFFF");
