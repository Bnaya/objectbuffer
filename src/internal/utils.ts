import { primitive, Entry } from "./interfaces";
import { ENTRY_TYPE } from "./entry-types";

const primitives = [
  "string",
  "number",
  "bigint",
  "boolean",
  "undefined"
] as const;

export function isPrimitive(value: unknown): value is primitive {
  if (primitives.includes(typeof value as any)) {
    return true;
  }

  if (value === null) {
    return true;
  }

  return false;
}

export function primitiveValueToEntry(value: primitive): Entry {
  if (typeof value === "string") {
    return {
      type: ENTRY_TYPE.STRING,
      value
    };
  }

  if (typeof value === "number") {
    return {
      type: ENTRY_TYPE.NUMBER,
      value
    };
  }

  if (typeof value === "bigint") {
    return {
      type: ENTRY_TYPE.BIGINT,
      value
    };
  }

  if (typeof value === "boolean") {
    return {
      type: ENTRY_TYPE.BOOLEAN,
      value
    };
  }

  if (typeof value === "undefined") {
    return {
      type: ENTRY_TYPE.UNDEFINED
    };
  }

  if (value === null) {
    return {
      type: ENTRY_TYPE.NULL
    };
  }

  throw new Error("unexpected");
}

export function createKnownTypeGuard<T>(arr: ReadonlyArray<T>) {
  return function knownTypeGuard(v: unknown): v is T {
    return arr.includes(v as any);
  };
}

export function invariant(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(message);
  }
}

export function arrayBufferCopyTo(
  origin: ArrayBuffer,
  startByte: number,
  length: number,
  target: ArrayBuffer,
  toTargetByte: number
) {
  const copyFrom = new Uint8Array(origin);
  const copyTo = new Uint8Array(target);

  for (let i = 0; i < length; i += 1) {
    copyTo[toTargetByte + i] = copyFrom[startByte + i];
  }
}
