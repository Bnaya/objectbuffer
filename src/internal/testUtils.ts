import { FIRST_FREE_BYTE_POINTER_TO_POINTER } from "./consts";

export function arrayBuffer2HexArray(
  buffer: ArrayBuffer,
  withByteNumber = false
) {
  if (withByteNumber) {
    return Array.prototype.map.call(
      new Uint8Array(buffer),
      (x: number, index) => `${index}:${x.toString(16).padStart(4, "0x")}`
    );
  }
  return Array.prototype.map.call(new Uint8Array(buffer), (x: number) =>
    x.toString(16).padStart(4, "0x")
  );
}

export function getFirstFreeByte(arrayBuffer: ArrayBuffer) {
  return new DataView(arrayBuffer).getUint32(
    FIRST_FREE_BYTE_POINTER_TO_POINTER
  );
}
