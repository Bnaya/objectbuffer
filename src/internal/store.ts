import { ENTRY_TYPE, isPrimitiveEntryType } from "./entry-types";
import { Entry, ObjectPropEntry, primitive } from "./interfaces";
import { arrayBufferCopyTo, isPrimitive, primitiveValueToEntry } from "./utils";
import { ExternalArgs } from "./interfaces";

export function initializeArrayBuffer(arrayBuffer: ArrayBuffer) {
  const dataView = new DataView(arrayBuffer);

  // global lock
  dataView.setInt32(0, 0);

  // End of data pointer / first free byte
  dataView.setUint32(8, 24);

  // first entry pointer
  dataView.setUint32(16, 24);

  return dataView;
}

export function writeEntry(
  externalArgs: ExternalArgs,
  dataView: DataView,
  startingCursor: number,
  entry: Entry
) {
  let cursor = startingCursor;

  // let writtenDataSizeInBytes = 0;

  // write type
  // undo on throw ?
  dataView.setUint8(cursor, entry.type);
  cursor += Uint8Array.BYTES_PER_ELEMENT;

  switch (entry.type) {
    case ENTRY_TYPE.UNDEFINED:
      break;

    case ENTRY_TYPE.NULL:
      break;

    case ENTRY_TYPE.BOOLEAN:
      dataView.setUint8(cursor, entry.value ? 1 : 0);
      cursor += Uint8Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.NUMBER:
      dataView.setFloat64(cursor, entry.value);
      cursor += Float64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.STRING:
      // eslint-disable-next-line no-case-declarations
      const encodedString: Uint8Array = externalArgs.textEncoder.encode(
        entry.value
      );
      dataView.setUint16(cursor, encodedString.byteLength);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      dataView.setUint16(cursor, entry.allocatedBytes);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      for (let i = 0; i < encodedString.length; i++) {
        dataView.setUint8(cursor, encodedString[i]);
        cursor += Uint8Array.BYTES_PER_ELEMENT;
      }

      cursor += entry.allocatedBytes;

      break;

    case ENTRY_TYPE.BIGINT:
      dataView.setBigInt64(cursor, entry.value);
      cursor += BigInt64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.UBIGINT:
      dataView.setBigUint64(cursor, entry.value);
      cursor += BigUint64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT:
      dataView.setUint32(cursor, entry.value);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT_PROP:
      // eslint-disable-next-line no-case-declarations
      const encodedStringKey: Uint8Array = externalArgs.textEncoder.encode(
        entry.value.key
      );
      dataView.setUint16(cursor, encodedStringKey.byteLength);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      for (let i = 0; i < encodedStringKey.length; i++) {
        dataView.setUint8(cursor, encodedStringKey[i]);
        cursor += Uint8Array.BYTES_PER_ELEMENT;
      }

      dataView.setUint32(cursor, entry.value.value);
      cursor += Uint32Array.BYTES_PER_ELEMENT;

      dataView.setUint32(cursor, entry.value.next);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.ARRAY:
      dataView.setUint32(cursor, entry.value);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      dataView.setUint32(cursor, entry.length);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      dataView.setUint32(cursor, entry.allocatedLength);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    default:
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore
      throw new Error(ENTRY_TYPE[entry.type] + " Not implemented yet");
  }

  return cursor - startingCursor;
}

export function appendEntry(
  externalArgs: ExternalArgs,
  dataView: DataView,
  entry: Entry
) {
  // End of data pointer
  const firstFreeByte = dataView.getUint32(8);

  const written = writeEntry(externalArgs, dataView, firstFreeByte, entry);
  dataView.setUint32(8, firstFreeByte + written);

  return {
    start: firstFreeByte,
    length: written
  };
}

export function readEntry(
  externalArgs: ExternalArgs,
  dataView: DataView,
  startingCursor: number
): [Entry, number] {
  let cursor = startingCursor;

  const entryType: ENTRY_TYPE = dataView.getUint8(cursor);
  cursor += Uint8Array.BYTES_PER_ELEMENT;

  const entry: any = {
    type: entryType,
    value: undefined as any
  };

  // let writtenDataSizeInBytes = 0;

  switch (entryType) {
    case ENTRY_TYPE.UNDEFINED:
      break;

    case ENTRY_TYPE.NULL:
      break;

    case ENTRY_TYPE.BOOLEAN:
      entry.value = dataView.getUint8(cursor) !== 0;
      cursor += Uint8Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.NUMBER:
      entry.value = dataView.getFloat64(cursor);
      cursor += Float64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.STRING:
      // eslint-disable-next-line no-case-declarations
      const stringLength = dataView.getUint16(cursor);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      entry.allocatedBytes = dataView.getUint16(cursor);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      // this wrapping is needed until:
      // https://github.com/whatwg/encoding/issues/172
      // eslint-disable-next-line no-case-declarations
      const tempAB = new ArrayBuffer(stringLength);
      arrayBufferCopyTo(dataView.buffer, cursor, stringLength, tempAB, 0);

      entry.value = externalArgs.textDecoder.decode(tempAB);

      cursor += stringLength;

      break;

    case ENTRY_TYPE.BIGINT:
      entry.value = dataView.getBigInt64(cursor);
      cursor += BigInt64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.UBIGINT:
      entry.value = dataView.getBigUint64(cursor);
      cursor += BigUint64Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT:
      entry.value = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    case ENTRY_TYPE.OBJECT_PROP:
      // eslint-disable-next-line no-case-declarations
      const keyStringLength = dataView.getUint16(cursor);
      cursor += Uint16Array.BYTES_PER_ELEMENT;

      // this wrapping is needed until:
      // https://github.com/whatwg/encoding/issues/172
      // eslint-disable-next-line no-case-declarations
      const tempAB2 = new ArrayBuffer(keyStringLength);
      arrayBufferCopyTo(dataView.buffer, cursor, keyStringLength, tempAB2, 0);

      // eslint-disable-next-line no-case-declarations
      const objectPropsValue: ObjectPropEntry["value"] = {
        key: externalArgs.textDecoder.decode(tempAB2),
        value: dataView.getUint32(cursor + keyStringLength),
        next: dataView.getUint32(
          cursor + keyStringLength + Uint32Array.BYTES_PER_ELEMENT
        )
      };

      cursor +=
        keyStringLength +
        Uint32Array.BYTES_PER_ELEMENT +
        Uint32Array.BYTES_PER_ELEMENT;

      entry.value = objectPropsValue;
      break;

    case ENTRY_TYPE.ARRAY:
      entry.value = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      entry.length = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      entry.allocatedLength = dataView.getUint32(cursor);
      cursor += Uint32Array.BYTES_PER_ELEMENT;
      break;

    default:
      throw new Error(ENTRY_TYPE[entryType] + " Not implemented yet");
  }

  return [entry, cursor - startingCursor];
}

export function reserveMemory(dataView: DataView, length: number) {
  const firstFreeByte = dataView.getUint32(8);

  dataView.setUint32(8, firstFreeByte + length);

  return firstFreeByte;
}

export function canSaveValueIntoEntry(
  externalArgs: ExternalArgs,
  entryA: Entry,
  value: primitive
) {
  const typeofTheValue = typeof value;

  if (entryA.type === ENTRY_TYPE.BOOLEAN && typeofTheValue === "boolean") {
    return true;
  }

  // number & bigint 64 are the same size
  if (
    (entryA.type === ENTRY_TYPE.BIGINT ||
      entryA.type === ENTRY_TYPE.UBIGINT ||
      entryA.type === ENTRY_TYPE.NUMBER) &&
    (typeofTheValue === "bigint" || typeofTheValue === "number")
  ) {
    return true;
  }

  if (
    entryA.type === ENTRY_TYPE.STRING &&
    typeofTheValue === "string" &&
    entryA.allocatedBytes >= externalArgs.textEncoder.encode(value).length
  ) {
    return true;
  }

  if (
    ((entryA.type === ENTRY_TYPE.NULL ||
      entryA.type === ENTRY_TYPE.UNDEFINED) &&
      value === undefined) ||
    value === null
  ) {
    return true;
  }

  return false;
}

export function overwriteEntryIfPossible(
  externalArgs: ExternalArgs,
  dataView: DataView,
  pointerToExistingEntry: number,
  value: any
) {
  if (isPrimitive(value)) {
    const existingValueEntry = readEntry(
      externalArgs,
      dataView,
      pointerToExistingEntry
    );

    if (canSaveValueIntoEntry(externalArgs, existingValueEntry[0], value)) {
      const stringAllocatedBytes =
        existingValueEntry[0].type === ENTRY_TYPE.STRING
          ? existingValueEntry[0].allocatedBytes
          : 0;

      const newEntry = primitiveValueToEntry(
        externalArgs,
        value,
        stringAllocatedBytes
      );

      writeEntry(externalArgs, dataView, pointerToExistingEntry, newEntry);
      return true;
    }

    return false;
  }

  return false;
}
