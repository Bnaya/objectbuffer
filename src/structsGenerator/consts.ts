import { fromEntries } from "../internal/utils";

export const typedArraysPropNameToCtorMap = {
  u8: Uint8Array,
  u8c: Uint8ClampedArray,
  i8: Int8Array,
  u16: Uint16Array,
  i16: Int16Array,
  u32: Uint32Array,
  i32: Int32Array,
  f32: Float32Array,
  f64: Float64Array,
  b64: BigInt64Array,
  u64: BigUint64Array,
} as const;

export type TypedArrayPropNameToCtorType = typeof typedArraysPropNameToCtorMap;

const intermediate1 = Object.entries(typedArraysPropNameToCtorMap).map(
  ([key, value]) => {
    return [value.name, key];
  }
);

export const typedArrayNameToHeapProp: any = fromEntries(intermediate1);

export type Heap = {
  [x in keyof TypedArrayPropNameToCtorType]: InstanceType<
    TypedArrayPropNameToCtorType[x]
  >;
};

export type PossibleTypedArrays =
  TypedArrayPropNameToCtorType[keyof TypedArrayPropNameToCtorType];
export type StructManifest = { [x: string]: PossibleTypedArrays };

export function createHeap(sb: ArrayBuffer | SharedArrayBuffer): Heap {
  const r = Object.fromEntries(
    Object.entries(typedArraysPropNameToCtorMap).map(([name, Ctor]) => [
      name,
      new Ctor(sb),
    ])
  );

  return r as any;
}
