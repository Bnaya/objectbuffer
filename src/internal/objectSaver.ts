import { appendEntry } from "./store";
import { ENTRY_TYPE } from "./entry-types";
import {
  ObjectEntry,
  ExternalArgs,
  GlobalCarrier,
  MapEntry,
  SetEntry
} from "./interfaces";
import { saveValue } from "./saveValue";
import { createHashMap, hashMapInsertUpdate } from "./hashmap/hashmap";

export function objectSaver(
  externalArgs: ExternalArgs,
  carrier: GlobalCarrier,
  referencedPointers: number[],
  objectToSave: any
) {
  const objectEntries = Object.entries(objectToSave);

  const hashMapPointer = createHashMap(
    carrier,
    Math.max(externalArgs.hashMapMinInitialCapacity, objectEntries.length * 1.3)
  );

  for (const [key, value] of objectEntries) {
    const ptrToPtr = hashMapInsertUpdate(
      externalArgs,
      carrier,
      hashMapPointer,
      key
    );

    const pointerToValue = saveValue(
      externalArgs,
      carrier,
      referencedPointers,
      value
    );

    carrier.dataView.setUint32(ptrToPtr, pointerToValue);
  }

  const objectStartEntry: ObjectEntry = {
    type: ENTRY_TYPE.OBJECT,
    refsCount: 1,
    value: hashMapPointer
  };

  return appendEntry(externalArgs, carrier, objectStartEntry);
}

export function mapSaver(
  externalArgs: ExternalArgs,
  carrier: GlobalCarrier,
  referencedPointers: number[],
  mapToSave: Map<string | number, any>
) {
  const hashMapPointer = createHashMap(
    carrier,
    Math.max(externalArgs.hashMapMinInitialCapacity, mapToSave.size * 1.3)
  );

  for (const [key, value] of mapToSave.entries()) {
    const ptrToPtr = hashMapInsertUpdate(
      externalArgs,
      carrier,
      hashMapPointer,
      key
    );

    const pointerToValue = saveValue(
      externalArgs,
      carrier,
      referencedPointers,
      value
    );

    carrier.dataView.setUint32(ptrToPtr, pointerToValue);
  }

  const objectStartEntry: MapEntry = {
    type: ENTRY_TYPE.MAP,
    refsCount: 1,
    value: hashMapPointer
  };

  return appendEntry(externalArgs, carrier, objectStartEntry);
}

export function setSaver(
  externalArgs: ExternalArgs,
  carrier: GlobalCarrier,
  setToSave: Set<string | number>
) {
  const hashMapPointer = createHashMap(
    carrier,
    Math.max(externalArgs.hashMapMinInitialCapacity, setToSave.size * 1.3)
  );

  for (const key of setToSave.keys()) {
    const ptrToPtr = hashMapInsertUpdate(
      externalArgs,
      carrier,
      hashMapPointer,
      key
    );

    carrier.dataView.setUint32(ptrToPtr, 1);
  }

  const objectStartEntry: SetEntry = {
    type: ENTRY_TYPE.SET,
    refsCount: 1,
    value: hashMapPointer
  };

  return appendEntry(externalArgs, carrier, objectStartEntry);
}
