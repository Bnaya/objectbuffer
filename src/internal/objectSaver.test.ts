/* eslint-env jest */

import { initializeArrayBuffer } from "./store";
import * as util from "util";
import { arrayBuffer2HexArray } from "./testUtils";
import { objectSaver } from "./objectSaver";
import { MemPool } from "@thi.ng/malloc";
import { MEM_POOL_START } from "./consts";
import { externalArgsApiToExternalArgsApi } from "./utils";

describe("objectSaver tests", () => {
  const externalArgs = externalArgsApiToExternalArgsApi({
    textEncoder: new util.TextEncoder(),
    textDecoder: new util.TextDecoder()
  });

  describe("objectSaver - general", () => {
    test("objectSaver", () => {
      const arrayBuffer = new ArrayBuffer(1024);
      const dataView = new DataView(arrayBuffer);
      initializeArrayBuffer(arrayBuffer);
      const allocator = new MemPool({
        buf: arrayBuffer,
        start: MEM_POOL_START
      });

      const objectToSave = {
        a: 6,
        b: "imastringa",
        6: null,
        ab: undefined,
        abc: BigInt("100"),
        nestedObject: {
          nestedProp: 7
        }
      };

      const saverOutput = objectSaver(
        externalArgs,
        { dataView, allocator },
        [],
        objectToSave
      );

      expect(saverOutput).toMatchInlineSnapshot(`800`);

      expect(arrayBuffer2HexArray(arrayBuffer, true)).toMatchSnapshot();
      expect(allocator.stats().available).toMatchInlineSnapshot(`216`);
    });
  });
});
