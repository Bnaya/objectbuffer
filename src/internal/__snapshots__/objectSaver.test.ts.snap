// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`objectSaver tests objectSaver - general objectSaver 2`] = `
Array [
  "0:0x00",
  "1:0x00",
  "2:0x00",
  "3:0x00",
  "4:0x00",
  "5:0x00",
  "6:0x00",
  "7:0x00",
  "8:0x00",
  "9:0x00",
  "10:0x00",
  "11:0xb7",
  "12:0x00",
  "13:0x00",
  "14:0x00",
  "15:0x00",
  "16:0x00",
  "17:0x00",
  "18:0x00",
  "19:0x18",
  "20:0x00",
  "21:0x00",
  "22:0x00",
  "23:0x00",
  "24:0x02",
  "25:0x40",
  "26:0x1c",
  "27:0x00",
  "28:0x00",
  "29:0x00",
  "30:0x00",
  "31:0x00",
  "32:0x00",
  "33:0x08",
  "34:0x00",
  "35:0x0a",
  "36:0x6e",
  "37:0x65",
  "38:0x73",
  "39:0x74",
  "40:0x65",
  "41:0x64",
  "42:0x50",
  "43:0x72",
  "44:0x6f",
  "45:0x70",
  "46:0x00",
  "47:0x00",
  "48:0x00",
  "49:0x18",
  "50:0x00",
  "51:0x00",
  "52:0x00",
  "53:0x00",
  "54:0x07",
  "55:0x00",
  "56:0x00",
  "57:0x00",
  "58:0x21",
  "59:0x08",
  "60:0x00",
  "61:0x0c",
  "62:0x6e",
  "63:0x65",
  "64:0x73",
  "65:0x74",
  "66:0x65",
  "67:0x64",
  "68:0x4f",
  "69:0x62",
  "70:0x6a",
  "71:0x65",
  "72:0x63",
  "73:0x74",
  "74:0x00",
  "75:0x00",
  "76:0x00",
  "77:0x36",
  "78:0x00",
  "79:0x00",
  "80:0x00",
  "81:0x00",
  "82:0x03",
  "83:0x00",
  "84:0x00",
  "85:0x00",
  "86:0x00",
  "87:0x00",
  "88:0x00",
  "89:0x00",
  "90:0x64",
  "91:0x08",
  "92:0x00",
  "93:0x03",
  "94:0x61",
  "95:0x62",
  "96:0x63",
  "97:0x00",
  "98:0x00",
  "99:0x00",
  "100:0x52",
  "101:0x00",
  "102:0x00",
  "103:0x00",
  "104:0x3b",
  "105:0x00",
  "106:0x08",
  "107:0x00",
  "108:0x02",
  "109:0x61",
  "110:0x62",
  "111:0x00",
  "112:0x00",
  "113:0x00",
  "114:0x69",
  "115:0x00",
  "116:0x00",
  "117:0x00",
  "118:0x5b",
  "119:0x05",
  "120:0x00",
  "121:0x0a",
  "122:0x69",
  "123:0x6d",
  "124:0x61",
  "125:0x73",
  "126:0x74",
  "127:0x72",
  "128:0x69",
  "129:0x6e",
  "130:0x67",
  "131:0x61",
  "132:0x08",
  "133:0x00",
  "134:0x01",
  "135:0x62",
  "136:0x00",
  "137:0x00",
  "138:0x00",
  "139:0x77",
  "140:0x00",
  "141:0x00",
  "142:0x00",
  "143:0x6a",
  "144:0x02",
  "145:0x40",
  "146:0x18",
  "147:0x00",
  "148:0x00",
  "149:0x00",
  "150:0x00",
  "151:0x00",
  "152:0x00",
  "153:0x08",
  "154:0x00",
  "155:0x01",
  "156:0x61",
  "157:0x00",
  "158:0x00",
  "159:0x00",
  "160:0x90",
  "161:0x00",
  "162:0x00",
  "163:0x00",
  "164:0x84",
  "165:0x01",
  "166:0x08",
  "167:0x00",
  "168:0x01",
  "169:0x36",
  "170:0x00",
  "171:0x00",
  "172:0x00",
  "173:0xa5",
  "174:0x00",
  "175:0x00",
  "176:0x00",
  "177:0x99",
  "178:0x07",
  "179:0x00",
  "180:0x00",
  "181:0x00",
  "182:0xa6",
  "183:0x00",
  "184:0x00",
  "185:0x00",
  "186:0x00",
  "187:0x00",
  "188:0x00",
  "189:0x00",
  "190:0x00",
  "191:0x00",
  "192:0x00",
  "193:0x00",
  "194:0x00",
  "195:0x00",
  "196:0x00",
  "197:0x00",
  "198:0x00",
  "199:0x00",
  "200:0x00",
  "201:0x00",
  "202:0x00",
  "203:0x00",
  "204:0x00",
  "205:0x00",
  "206:0x00",
  "207:0x00",
  "208:0x00",
  "209:0x00",
  "210:0x00",
  "211:0x00",
  "212:0x00",
  "213:0x00",
  "214:0x00",
  "215:0x00",
  "216:0x00",
  "217:0x00",
  "218:0x00",
  "219:0x00",
  "220:0x00",
  "221:0x00",
  "222:0x00",
  "223:0x00",
  "224:0x00",
  "225:0x00",
  "226:0x00",
  "227:0x00",
  "228:0x00",
  "229:0x00",
  "230:0x00",
  "231:0x00",
  "232:0x00",
  "233:0x00",
  "234:0x00",
  "235:0x00",
  "236:0x00",
  "237:0x00",
  "238:0x00",
  "239:0x00",
  "240:0x00",
  "241:0x00",
  "242:0x00",
  "243:0x00",
  "244:0x00",
  "245:0x00",
  "246:0x00",
  "247:0x00",
  "248:0x00",
  "249:0x00",
  "250:0x00",
  "251:0x00",
  "252:0x00",
  "253:0x00",
  "254:0x00",
  "255:0x00",
]
`;
